### DISCUSSION TITLE
Storage of information for the customisation feature

### DISCUSSION THREAD

**Discussion by joulev** [joulev/ezkomment#196] [General]
Before, I thought of storing the user-provided source directly in the database.

That is definitely an option to consider, but today I figure it may be better to store the patches instead, to resemble how `git` works.

```sh
# to add it to package.json; it is actually already installed (as dependency of a different package)
yarn add diff
```

```js
const Diff = require("diff");

const before = `const fs = require("fs");
const path = require("path");
const utils = require("../utils");\n`;
const after = `const fs = require("fs");
const path = require("path");

console.log(process.cwd());\n`;

const patch = Diff.createPatch("test.js", before, after); // store this

console.log(patch);
```

![](https://user-images.githubusercontent.com/44609036/178231107-021f012d-44e8-4a08-b441-e4a48181dd34.png)

Then when we need any version of the HTML, we simply need to chain `Diff.applyPatch()` from the default HTML until the revision we need.

Pros: Allow users to store all deployments, and can deploy any revisions they want. For example in Firebase I can access old versions of the security rules any time

<img width="1182" alt="image" src="https://user-images.githubusercontent.com/44609036/178231810-835aa4bb-9ace-436b-a6aa-03f9d50b8ddb.png">

Cons: Slower (of course) and more things stored in the database.

In any case this will require a new route (we should not combine this to `GET /sites/:id`) because it is a pretty big performance penalty.

**Comment by VietAnh1010** [comment]
> In any case this will require a new route (we should not combine this to GET /sites/:id) because it is a pretty big performance penalty.

Agree.

About the feature, I am not sure about this at the moment. Is it really necessary to store different versions of the configuration file?

I will first create an endpoint to upload and store the configuration file. Should I store it in Firestore, or in Storage?

**Reply by joulev** [reply]
> Is it really necessary to store different versions of the configuration file?

It's exactly the topic of this discussion.

> Should I store it in Firestore, or in Storage?

Firestore since it won't actually be a file; we will only save it as a string verbatim.

**Reply by joulev** [reply]
For simplicity's sake, let's not save all revisions for now. Only the last version of the file is saved just like any normal other user information field. "Save" button removed (5dc86dc7895caff5a38810f1e030236341ddeb6b).

I think later on if we ever want to implement the revisions instead it should not be too complicated (it may not even be a breaking change).

**Reply by VietAnh1010** [reply]
Agree

**Reply by joulev** [reply]
Let's start working on this, we are 11 days away from MS3 so better start early.

**Reply by VietAnh1010** [reply]
Completed the initial implementation. Currently, we will store the configuration as a string.
